### ISSUE TITLE
Auto-deploy main to both dev and staging envs

### ISSUE THREAD

**Issue by Bento007** [chanzuckerberg/single-cell-data-portal#5461]
This involves having the main branch GHA deploy to staging (in addition to dev, which requires no change). Practically, this mean that upon merge to main, merge main to staging branch.

This applies to both 
- [ ] ~https://github.com/chanzuckerberg/single-cell-explorer~
- [ ] https://github.com/chanzuckerberg/single-cell-data-portal

**Comment by Bento007** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/5461?planning-poker) with Zenhub @tihuan

**Comment by Bento007** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/5461?planning-poker) with Zenhub @ebezzi

**Comment by prathapsridharan** [comment]
@atolopko-czi @tihuan (cc: @Bento007 ) - In our meeting on 09/28/2023 we agreed that to enable a `staging` deploy when a commit is pushed to `main`, we can leverage our current workflow by merging the `main` branch into `staging` branch which would [create a deployment on a merge to staging branch](https://github.com/chanzuckerberg/single-cell-data-portal/blob/d042bdadb17f2fa389f537ed26eb429698350596/.github/workflows/build-images-and-create-deployment.yml#L80) and then [kick off the creation of staging stack](https://github.com/chanzuckerberg/single-cell-data-portal/blob/main/.github/workflows/deploy-happy-stack.yml)

This is all well and good. But I have a couple of questions:

_What happens if there is a burst of commits to `main`?_ Per above, we would:

```
git checkout staging
git merge --no-ff main -m "Merge main into staging"
git push
```

Each of these merges to `staging` would trigger a creation of a staging deployment and creation of a stack. 

1. _Would these workloads just get queued up?_ This would mean stack creation could take a long time. I think this problem is no worse than it is currently for `dev` deployment when there is a burst of commits to `main` branch but just want to check with y'all.

2. _Is it possible for stacks to be created out of order due to a race condition?_ That is `commit A` _precedes_ `commit B` and _what you expect to see in staging is a build containing `commit B`_, but due to a race condition of some sort, `commit B` is processed first and then `commit A` is processed and what you have is `commit A` on staging.

3. Could the `git merge main` into `staging` run into a merge conflict scenario if there is a chance of out-of-order processing as stated in (2) above? If so, is it better to recreate the `staging` branch with the `main` branch? If we do that what would happen to queue of deployment workloads in (1) above?

I think (2) and (3) are also not problems but I am less sure here. I imagine, that merging to `staging` is always in-order (guaranteed?), and so our problems is reduced to the same situation as a `dev` deployment dealing with a burst of commits to `main` in that if race conditions on executing a build and deploy workflow is not an observed issue in `dev`, it won't be an issue in `staging`. But again I want to check with y'all here

**Comment by atolopko-czi** [comment]
While we had talked about reinstating merge queues to handle the ordering/race condition issue, what about using [this solution](https://stackoverflow.com/questions/66335225/how-to-cancel-previous-runs-in-the-pr-when-you-push-new-commitsupdate-the-curre)?

**Comment by prathapsridharan** [comment]
If we had a queue that will _guarantee_ all `commit A` workflows will run


> While we had talked about reinstating merge queues to handle the ordering/race condition issue, what about using [this solution](https://stackoverflow.com/questions/66335225/how-to-cancel-previous-runs-in-the-pr-when-you-push-new-commitsupdate-the-curre)?

That sounds good. I will add this.

However, it is worth noting that the potential for this problem currently exists in `dev` deployments - that is, several commits come to the `main` branch kicking off several GHA deployment workflows and theoretically, there is the possibility of a race condition such that an older commit deployment wins over the deployment of the latest commit.

**Comment by prathapsridharan** [comment]
I think this task has proven more complex that anticipated. The issues:

1. Branch protection of required `staging` branch requiring passing GHA prevented auto-merging from `main` to `staging`
2. After addressing (1) by removing GHA passing requirement in `staging` and successfully auto-merging `main` to `staging`, the expected behavior was that it would trigger subsequent workflow to trigger build and deploy workflows. That did not happen. Seems there github has [protection against workflows triggering other workflows](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow) even if an "on-trigger" criteria is met.

(1) and (2) make implementation and testing (particularly using test branches) quite difficult. One obvious (and painful) way to test this would be to make commits to `main` until we reach a state where a `stage` deploy is created. The other is to setup 2 test branches with all the protections and features of `main` and `staging` but that seems impractical.

That is just the essential complexity. The accidental complexity is introduced when we have other variables like flakey E2E tests, and needing to freeze environment for pen-testing, etc which is what transpired on the week of 10/09/2023.

Assuming we are able to address the above and get the pieces hooked up such that a merge from `main` to `staging` does happen on every commit to `main` and subsequently triggers a `stage` env deploy, we should also consider a couple of more things about what the state of `staging` branch is going to be:

A) _The ideal setup_ (I believe) is that `staging` branch is protected by _the requirement that only all unit tests in the incoming branch (ex: `main`) passes_. But that makes implementation even more complex in that a merge from `main` to `staging` should only be triggered when the unit tests passes and if that doesn't happen must fail the merge and send a slack notification.

B) We might take the more expedient route and state that the MVP setup is to remove all branch protection on `staging` and simply allow all commits to `main` to be merged into `staging`. This relies on merges from `rdev` (or PRs) into `main` such that all tests pass on `rdev`. Particularly we are relying on 2 separate `rdev` (PRs) whose tests are passing independently don't break `main` when both are merged.

**Comment by atolopko-czi** [comment]
Maybe we should reconsider deploying to staging env directly from `main` branch? This would eliminate having to make the main->staging merge work. (Now that we know about the `concurrency` feature of GHA, I think our earlier concerns about deployment race conditions are addressed.) Moving to an all-in-one workflow for merges to main might simplify things. E.g. a workflow that looks like:

```
Run unit tests on main +---+
                           |-- Deploy to staging --- Run E2E/Func tests
Build images --------------+
```

This would replace all of:
* push-tests.yml (would keep for use with branches)
* build-image-and-create-deployment.yml
* deploy-happy-stack.yml

Some related benefits to this approach:
* It would address https://github.com/chanzuckerberg/single-cell-data-portal/issues/5939 (i.e. fail fast if unit tests fail)
* Images would be built optimistically, but if unit tests fail, it aborts deployment
* We could view the entire CI/CD GHA task graph in one workflow

I may be missing good reasons why a single workflow is problematic, but at first blush it seems overall simpler to me.

**Comment by tihuan** [comment]
The proposal Andrew has above sounds good! The only thing is probably tracking the latest commit on `staging`, so oncall engineer knows which commit to use for prod deployment? The `concurrency` feature definitely solves lots of issues!

**Comment by prathapsridharan** [comment]
@atolopko-czi @tihuan @danieljhegeman (cc: @ebezzi @Bento007)

We should discuss in-person about several things and I am writing the following down to serve our discussion:

1. In principle an all-in-one workflow seems desirable. We are effectively copying the bulk of `push-tests.yml`, `build-images-and-create-deployment.yml`, and `deploy-happy-stack.yml` into a single workflow file with above _job dependency graph_. This basically amounts to a pretty big refactoring job _and we don't have good testing support_. 

_Doesn't this sounds like quite a bit of work? If so, I think it is important to critically assess how much work it is given that we expected to be wrapped up with CI/CD by now. This is not to say that we shouldn't do it but I think we should acknowledge the complexity and work required. The following points might help determine how much work this new approach is_

2. _How to test this new unified workflow?_ I don't think we want to disrupt the current execution of GHAs in `main` while this work is in progress because it will block engineers. _So we should have a mechanism to either fully test this and verify it is working before merging to `main` or have this workflow running in parallel to existing GHAs in `main`_.

Would running in parallel this work-in-progress workflow cause resource starvation problems on GHA servers that make existing GHAs on `main` slower (or worse, fail) and slow down development on `main`?

If running the work-in-progress workflow in parallel is not an option for the above reason, what is a scheme to fully test this unified workflow end-to-end before merging to `main`? Would we have a feature branch run this entire single workflow _including deploying to `stage` env there by clobbering `stage` env while work is in progress?_

4. What was the original reason to have separate workflow files? For readability?  For parallelism? 

If it is readability, moving to a single workflow file certainly affects readability albeit that is a minor concern compared to everything else.

_If it is parallelism, what if this single workflow architecture takes an unacceptable amount of time? We might then have to think hard about parallelizing again._

5. Is going with monolithic workflow rearchitecture a _currently_ better investment (given our current resource and time constraints) than getting our current scheme to work by figuring out [how to trigger a workflow from a workflow](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow)? That is, now that we can auto-merge to `staging` branch, figure out how to let this trigger the `stage` deploy.

6. Given the complexities stated above, should we consider lowering the priority on #5482 and #5479 since our current manual schemes are sufficient? Getting continuous deployment to `stage` env alone would yield significant value and it seems there is still quite a bit of work to do to get it working.

**Comment by tihuan** [comment]
Really great points too, Prathap! I think if we could figure out workflow triggering workflow, it's probably the lowest hanging fruit. It sounds like GitHub just wants to make sure people don't accidentally create a recursive loop of GitHub actions and incurring crazy amount of costs, but the foot gun is available for us, we just need to be v careful. Even better if there's a way to do cycle detection 😆

**Comment by prathapsridharan** [comment]
@atolopko-czi @tihuan @Bento007 - _I don't think using `concurrency` feature of GHA is a good thing here._ The main reason being that _[ordering is not guaranteed for jobs or runs using concurrency groups, they are handled in the order that they are processed](https://docs.github.com/en/actions/using-jobs/using-concurrency#overview)_.

Here is the pertinent parts of our workflow DAG:

<img width="1127" alt="Screenshot 2023-10-18 at 4 32 00 PM" src="https://github.com/chanzuckerberg/single-cell-data-portal/assets/5581819/0fc4b9a7-5557-43aa-9ff1-fac3a3b683e7">

You could have all kinds of scenarios where workflows get cancelled in undesirable ways depending on which workflows have the `concurrency` groups. Moreover, if we change the structure of the DAG, say push to `prod` should trigger `push-tests.yml`, then that could cause further confusion.

_I think what we need is ordering._

concurrency group makes sense when the collection of steps being executed are idempotent (ex: set x = 5). In this case, you don't need multiple tasks running `set x = 5`. But this is not the case for us. Each commit that triggers the workflow is an input to the workflow and that might change the success/failure status of the workflow. We need to capture all those workflow outputs to reason about things.
